--- linux/drivers/peci/controller/peci-aspeed.c	2025-02-26 10:05:36.088058999 +0800
+++ linux.new/drivers/peci/controller/peci-aspeed.c	2025-02-26 13:01:39.000000000 +0800
@@ -221,9 +221,18 @@
 
 	/* Check command sts and bus idle state */
 	ret = aspeed_peci_check_idle(priv);
-	if (ret)
+	if (ret) {
+		reset_control_assert(priv->rst);
+		reset_control_deassert(priv->rst);
+		aspeed_peci_init_regs(priv);
+		ret = clk_set_rate(priv->clk, priv->clk_frequency);
+		if (ret < 0) {
+		    dev_err(priv->dev, "Failed to set PECI clock frequency\n");
+		    return ret;
+		}
+    	aspeed_peci_controller_enable(priv);
 		return ret; /* -ETIMEDOUT */
-
+	}
 	spin_lock_irq(&priv->lock);
 	reinit_completion(&priv->xfer_complete);
 
